### ISSUE TITLE
EXC_BAD_ACCESS: Attempted to dereference garbage pointer

### ISSUE THREAD

**Issue by andeeliao** [BendingSpoons/tempura-swift#100]
The issue is a dereferenced pointer crash that happens occasionally when a navigation transition (occurring from a button tap that triggers `.hide(): .pop` happens at the same time as a state update (originating from a side effect).

Xcode version: 12.1
Target: iOS 12.2
Stack Trace: 

```
OS Version: iOS 13.5.1 (17F80)
Report Version: 104

Exception Type: EXC_BAD_ACCESS (SIGBUS)
Exception Codes: BUS_NOOP at 0x0000000115200008
Crashed Thread: 0

Application Specific Information:
18837 >
Attempted to dereference garbage pointer 0x115200008.

Thread 0 Crashed:
0   libswiftCore.dylib              0x3517d6b44         swift_release
1   OurApp                          0x200cbf0f0         AppState
2   Tempura                         0x1020e3d28         ViewController.storeDidChange (ViewController.swift:280)
3   Tempura                         0x1020d891c         mainThread (MainThread.swift:16)
4   Tempura                         0x1020e4c08         [inlined] ViewController.storeDidChange (ViewController.swift:279)
5   Tempura                         0x1020e4c08         [inlined] ViewController.subscribe (ViewController.swift:267)
6   Tempura                         0x1020e4c08         ViewController.subscribe
7   Katana                          0x286a37efc         [inlined] @callee_guaranteed
8   Katana                          0x286a37efc         @callee_guaranteed
9   Katana                          0x286a36540         [inlined] @callee_guaranteed
10  Katana                          0x286a36540         [inlined] Store.manageAction (Store.swift:537)
11  Katana                          0x286a36540         [inlined] $sIeg_s5Error_pIggzo_xlyytIsegr_sAA_pIegnzo_TR
12  Katana                          0x286a36540         [inlined] Sequence.forEach
13  Katana                          0x286a36540         Store.manageAction
14  Katana                          0x286a336d8         @callee_guaranteed
15  libdispatch.dylib               0x314eb49a4         _dispatch_call_block_and_release
16  libdispatch.dylib               0x314eb5520         _dispatch_client_callout
17  libdispatch.dylib               0x314e675b0         _dispatch_main_queue_callback_4CF$VARIANT$mp
18  CoreFoundation                  0x338e167f8         __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__
19  CoreFoundation                  0x338e116cc         __CFRunLoopRun
20  CoreFoundation                  0x338e10ce4         CFRunLoopRunSpecific
21  GraphicsServices                0x31b45f388         GSEventRunModal
22  UIKitCore                       0x33c057440         UIApplicationMain
23  OurApp                          0x200b457b0         main (AppDelegate.swift:11)
24  libdyld.dylib                   0x18e1bb8ec         start
```

We suspect in `ViewController`: line 280 `self.update(with: self.state) the ViewController (`self`) has been deallocated (we confirmed `deinit` is called from a background thread), but after `storeDidChange` is called (and queues up the update on the mainThread). Hence, when `self.update` is actually run on the mainThread, self has been deallocated. 

We've tried to disconnect before dispatching a `Hide` action (see below), but that doesn't seem to resolve the issue. (possibly because the disconnect still triggers a state update that creates the same race condition with the deallocation?)

```
self.connected = false
self.dispatch(Hide(animated: true))
```

Would this be resolved by adding a `[weak self]` to the mainThread call? A state update shouldn't be required if the ViewController is deallocated at any point.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1199108774605801/1199216780360977) by [Unito](https://www.unito.io/learn-more)

**Comment by smeador** [comment]
Support here would be greatly appreciated. Again, we've pinpointed the crash to this line of code in ViewController.swift: 280
```
    mainThread {
     self.update(with: self.state)
    }
```
We're running Tempura 4.4.0, but are in the process of upgrading. However looking at 6.1.0 this function is the same.

**Comment by DannyNemer** [comment]
My team has encountered the same bug.

**Comment by danyf90** [comment]
This has been fixed in Tempura 6.2.0, thanks for reporting!
